pyacq/core/nodegroup.py

```diff
@@ -1,50 +1,31 @@
 from pyqtgraph.Qt import QtCore, QtGui
 
 from .rpc import RPCServer
-from .node import Node, WidgetNode
+from .node import Node, WidgetNode, register_node_type
 from .nodelist import all_nodes
 
+import time
+import importlib
+import pickle
 
-
-class RpcThread( QtCore.QThread):
+class ThreadReadSocket( QtCore.QThread):
+    new_message = QtCore.Signal(QtCore.QByteArray, QtCore.QByteArray)
     def __init__(self, rpc_server, parent = None):
         QtCore.QThread.__init__(self, parent)
         self.rpc_server = rpc_server
-    
-    def run(self):
-        while self.rpc_server.running():
-            self.rpc_server._process_one()
-
-class NodeGroupApplication(QtGui.QApplication):
-    def create_widget_of_node(self):
-        node = self.sender()
-        node.create_widget()
         
-    def show_widget_of_node(self):
-        node = self.sender()
-        node.widget.show()
-
-
-class NodeGroup:
-    """
-    Node group is not directly a RPCServer.
-    """
-    def __init__(self, name, addr):
-        self.rpc_server = _NodeGroup(name, addr)
-        self._name = name
-        self._addr = self.rpc_server._addr
-        self.nodes = {}
-
-    def run_forever(self):
-        self.app = NodeGroupApplication([])
-        self.rpc_thread = RpcThread(self.rpc_server, parent = None)
-        self.rpc_thread.finished.connect(self.app.quit)
-        self.rpc_thread.start()
-        self.app.exec_()
     
+    def run(self):
+        self.go = True
+        while self.go:
+            # we need to poll in case of RPCServer.close()
+            event = self.rpc_server._socket.poll(500)
```

Comment:
For (3) :I do not like so much QTImer. because message are not executed immediatly but every XXms.
It is less responssive.
For (4): we need the NodeGroup to be in the main thread because NodeGroup.nodes['mynode'].method1() need to be call safely. A Node can be a widget. I made some test on window and linux and it works.

For the solution is :+1: 
   * recv and send in a thread or good mutex.
   * fn() in the main

No solution (1) was:
   - a thread is wating in background on recv (no poll).
   - when a message come the thread emit a message and stop.
   - the main thread process the message : exec fn() + send back result
   - the main thread start the thread again.
It was not so bad but start/stop overhead.